File: wavexlr/health.py
"""Observe audio health by default; disruptive remedies require explicit opt-in.

No graph-driver changes or profile preferences are imposed. stop() cancels
collection and waits for restoration of any transaction already started.
"""

import json
import logging
import re
import threading
import time

from . import recovery

log = logging.getLogger("wavexlr.health")
CHECK_INTERVAL = 10.0
GLITCH_XRUNS_PER_CHECK = 50
GLITCH_CONFIRM_CHECKS = 2
COOLDOWN_SECONDS = 60.0
MAX_ATTEMPTS = 2
STALL_CLEAN_REFILL_CHECKS = 6


def sample_xruns(runner=None):
    runner = runner or recovery.CommandRunner()
    text = runner.run(["pw-top", "--batch-mode", "--iterations", "3"], timeout=15)
    return _parse_pw_top(text) if text is not None else {}


def _parse_pw_top(text):
    counts = {}
    for line in text.splitlines():
        tokens = line.split()
        # S ID QUANT RATE WAIT BUSY W/Q B/Q ERR [FORMAT...] NAME
        if len(tokens) < 10 or not tokens[8].isdigit():
            continue
        # Later iterations overwrite earlier ones: last wins.
        counts[tokens[-1]] = int(tokens[8])
    return counts


def read_playback_status(card, device, subdevice):
    """(hw_ptr, state) of one ALSA playback substream, or (None, None).

    hw_ptr is the DMA position the hardware has consumed up to. A
    running playback stream always advances it — even one playing pure
    silence — which is what makes a static pointer a hardware verdict
    rather than a signal-level one.
    """
    path = (f"/proc/asound/card{card}/pcm{device}p/"
            f"sub{subdevice}/status")
    try:
        with open(path) as f:
            text = f.read()
    except OSError:
        return None, None
    ptr = re.search(r"^hw_ptr\s*:\s*(\d+)", text, re.MULTILINE)
    state = re.search(r"^state:\s*(\S+)", text, re.MULTILINE)
    return (int(ptr.group(1)) if ptr else None,
            state.group(1) if state else None)


def sample_source_mutes(runner=None):
    """Unknown mute state cannot enable recovery."""
    sources = recovery._listing("sources", runner)
    return {source["name"]: source["mute"] for source in sources
            if isinstance(source.get("name"), str) and isinstance(source.get("mute"), bool)}


def recycle_sink(sink_name, runner=None):
    """Resume even if cancellation or timeout interrupts suspension."""
    if runner and runner.stopped.is_set():
        return False
    suspended = False
    try:
        suspended = recovery._pactl("suspend-sink", sink_name, "1") is not None
        if suspended:
            (runner.stopped if runner else threading.Event()).wait(1.0)
    finally:
        resumed = recovery._pactl("suspend-sink", sink_name, "0") is not None
        if not resumed:
            log.error("Could not resume %s after suspension", sink_name)
    return suspended and resumed


def snapshot_graph(runner=None):
    """Return {capture: running}, watched sinks; None means unknown graph."""
    from .audio import SOURCE_MATCHES
    runner = runner or recovery.CommandRunner()
    text = runner.run(["pw-dump"])
    try:
        dump = json.loads(text)
    except (ValueError, TypeError):
        return None
    if not isinstance(dump, list):
        return None
    nodes = {}
    names = {}
    targets = set()
    captures = {}
    for obj in dump:
        if not isinstance(obj, dict) or obj.get("type") != "PipeWire:Interface:Node":
            continue
        info = obj.get("info") or {}
        props = info.get("props") or {}
        name = props.get("node.name", "")
        if not isinstance(name, str):
            continue
        names[obj["id"]] = name
        nodes[name] = (info, props)
        if props.get("media.class") == "Audio/Source" and name.startswith(SOURCE_MATCHES):
            captures[name] = info.get("state") == "running"
    for obj in dump:
        if not isinstance(obj, dict) or obj.get("type") != "PipeWire:Interface:Link":
            continue
        info = obj.get("info") or {}
        source = names.get(info.get("output-node-id"), "")
        target = names.get(info.get("input-node-id"), "")
        if source.startswith("openwave_loop_output_") and not source.endswith("_cap"):
            if target.startswith("alsa_output."):
                targets.add(target)
    sinks = {}
    for name in targets:
        if name not in nodes:
            continue
        info, props = nodes[name]
        if props.get("media.class") != "Audio/Sink":
            continue
        try:
            sinks[name] = {
                "running": info.get("state") == "running",
                "card": int(props.get("api.alsa.pcm.card", props.get("alsa.card"))),
                "device": int(props.get("api.alsa.pcm.device", props.get("alsa.device"))),
                "subdevice": int(props.get("api.alsa.pcm.subdevice", props.get("alsa.subdevice", 0))),
            }
        except (KeyError, TypeError, ValueError):
            continue
    return captures, sinks


class GlitchWatch:
    """Confirm sustained xrun growth; baselines are not healthy observations."""

    def __init__(self, threshold=GLITCH_XRUNS_PER_CHECK, confirm=GLITCH_CONFIRM_CHECKS):
        self.threshold = threshold
        self.confirm = confirm
        self._prev = {}
        self._streak = {}
        self._delta = {}

    def pause(self, node_name):
        for state in (self._prev, self._streak, self._delta):
            state.pop(node_name, None)

    def observe(self, node_name, xruns):
        prev = self._prev.get(node_name)
        self._prev[node_name] = xruns
        if prev is None or xruns < prev:
            self._streak[node_name] = 0
            self._delta[node_name] = None
            return False
        delta = xruns - prev
        self._delta[node_name] = delta
        self._streak[node_name] = self._streak.get(node_name, 0) + 1 if delta >= self.threshold else 0
        return delta >= self.threshold

    def glitching(self, node_name):
        return self._streak.get(node_name, 0) >= self.confirm

    def just_confirmed(self, node_name):
        return self._streak.get(node_name, 0) == self.confirm

    def last_delta(self, node_name):
        return self._delta.get(node_name)


class SinkStallWatch:
    """Decides when a running sink's hardware has stopped consuming.

    Fed (running, hw_ptr, alsa_state) per check window. A stall is a
    pointer that did not move between two windows while the node claims
    to be running, or the kernel reporting the stream in XRUN. The first
    observation of a sink only baselines the pointer — a sink that just
    started gets a full window before being judged.
    """

    def __init__(self, cooldown_seconds=COOLDOWN_SECONDS,
                 max_attempts=MAX_ATTEMPTS,
                 clean_refill=STALL_CLEAN_REFILL_CHECKS):
        self.cooldown_seconds = cooldown_seconds
        self.max_attempts = max_attempts
        self.clean_refill = clean_refill
        self._prev_ptr = {}      # sink_name -> last hw_ptr
        self._stalled = {}       # sink_name -> bool
        self._was_stalled = {}   # sink_name -> stalled on previous window
        self._clean = {}         # sink_name -> consecutive moving windows
        self._attempts = {}      # sink_name -> remedies spent
        self._last_attempt = {}  # sink_name -> monotonic time

    def forget(self, sink_name):
        for d in (self._prev_ptr, self._stalled, self._was_stalled,
                  self._clean, self._attempts, self._last_attempt):
            d.pop(sink_name, None)

    def observe(self, sink_name, running, hw_ptr, alsa_state, now):
        """Account one window; True when the sink is stalled."""
        self._was_stalled[sink_name] = self._stalled.get(sink_name, False)
        prev = self._prev_ptr.get(sink_name)
        self._prev_ptr[sink_name] = hw_ptr
        if not running or hw_ptr is None or alsa_state not in ("RUNNING", "XRUN"):
            # Idle and suspended sinks legitimately hold still, and a
            # sink whose /proc entry vanished is not ours to judge.
            self._stalled[sink_name] = False
            self._prev_ptr.pop(sink_name, None)
            self._clean[sink_name] = 0
            return False
        if alsa_state == "XRUN":
            self._stalled[sink_name] = True
            self._clean[sink_name] = 0
            return True
        if prev is None:
            self._stalled[sink_name] = False
            return False
        stalled = hw_ptr == prev
        self._stalled[sink_name] = stalled
        if stalled:
            self._clean[sink_name] = 0
        else:
            # Same reasoning as the glitch watch, shorter leash: a
            # recycle resets the pointer and the next window can move
            # once without the PCM being healthy, so refill only after
            # a sustained stretch of movement.
            clean = self._clean.get(sink_name, 0) + 1
            self._clean[sink_name] = clean
            if clean >= self.clean_refill:
                self._attempts.pop(sink_name, None)
        return stalled

    def just_stalled(self, sink_name):
        """True on the window a stall begins, for logging it once."""
        return (self._stalled.get(sink_name, False)
                and not self._was_stalled.get(sink_name, False))

    def spent(self, sink_name):
        """Remedy attempts spent on the current incident."""
        return self._attempts.get(sink_name, 0)

    def should_recover(self, sink_name, now):
        if not self._stalled.get(sink_name):
            return False
        if self._attempts.get(sink_name, 0) >= self.max_attempts:
            return False
        last = self._last_attempt.get(sink_name)
        if last is not None and now - last < self.cooldown_seconds:
            return False
        return True

    def record_attempt(self, sink_name, now):
        self._attempts[sink_name] = self._attempts.get(sink_name, 0) + 1
        self._last_attempt[sink_name] = now
        # The recycle itself resets the pointer; don't let the next
        # window compare against a pre-recycle value.
        self._prev_ptr.pop(sink_name, None)


class HealthMonitor:
    """Log confirmed faults; only auto_recover=True permits remedies."""

    def __init__(self, auto_recover=False, capture_gaps=None):
        self.auto_recover = auto_recover
        self._capture_gaps = capture_gaps
        self._stop = threading.Event()
        self._runner = recovery.CommandRunner(self._stop)
        self._thread = None
        self._lifecycle = threading.Lock()
        self._check_lock = threading.Lock()
        self.glitch = GlitchWatch()
        self.capture = recovery.StallWatch()
        self._no_data = set()
        self.stall = SinkStallWatch()
        self._known_captures = set()
        self._known_sinks = set()

    def start(self):
        with self._lifecycle:
            if self._thread is not None and self._thread.is_alive():
                return
            self._stop.clear()
            self._thread = threading.Thread(target=self._run, daemon=True,
                                            name="openwave-health")
            self._thread.start()

    def stop(self):
        """Cancel/reap collection and wait for any owed restoration.

        No timed join: a fifteen-second sampler cannot outlive shutdown.
        Also waits for callers driving check_once directly.
        """
        with self._lifecycle:
            self._runner.cancel()
            if self._thread is not None:
                self._thread.join()
                self._thread = None
            with self._check_lock:
                pass

    def check_once(self, now=None):
        with self._check_lock:
            if self._stop.is_set():
                return
            self._check_once(now)

    def _pause_capture(self, name):
        self.glitch.pause(name)
        self.capture.pause(name)
        self._no_data.discard(name)

    def _check_once(self, now):
        graph = snapshot_graph(self._runner)
        if graph is None or self._stop.is_set():
            for name in self._known_captures:
                self._pause_capture(name)
            for name in self._known_sinks:
                self.stall.observe(name, False, None, None, now)
            return
        captures, sinks = graph
        for gone in self._known_captures - set(captures):
            # Profile cycling and graph reconstruction can temporarily remove
            # the same node. Absence is not proof that its incident ended.
            self._pause_capture(gone)
        for gone in self._known_sinks - set(sinks):
            self.stall.observe(gone, False, None, None, now)
        self._known_captures = set(captures)
        self._known_sinks = set(sinks)
        if captures:
            counts = sample_xruns(self._runner)
            mutes = sample_source_mutes(self._runner)
            gaps = self._capture_gaps() if self._capture_gaps is not None else {}
            observed_at = time.monotonic() if now is None else now
            for name, running in captures.items():
                if self._stop.is_set():
                    return
                if not running or mutes.get(name) is not False:
                    self._pause_capture(name)
                    continue
                if name in counts:
                    self.glitch.observe(name, counts[name])
                else:
                    self.glitch.pause(name)
                age = gaps.get(name)
                no_data = age is not None and age >= self.capture.stall_seconds
                if no_data and name not in self._no_data:
                    log.warning("%s has no capture data for %.1fs; auto-recover %s",
                                name, age, self.auto_recover)
                if no_data:
                    self._no_data.add(name)
                else:
                    self._no_data.discard(name)
                delta = self.glitch.last_delta(name)
                healthy = (delta is not None and delta < self.glitch.threshold
                           and (self._capture_gaps is None
                                or (age is not None and age < self.capture.stall_seconds)))
                if healthy:
                    self.capture.record_recovered(name, observed_at)
                else:
                    self.capture.pause(name)
                if self.glitch.just_confirmed(name):
                    log.warning("%s has sustained capture xruns (%d/window); auto-recover %s",
                                name, self.glitch.last_delta(name), self.auto_recover)
                fault = no_data or self.glitch.glitching(name)
                if self.auto_recover and fault and self.capture.can_recover(name, observed_at):
                    card = recovery.card_name_for(name, self._runner)
                    if card and not self._stop.is_set():
                        self.capture.record_attempt(name, time.monotonic() if now is None else now)
                        recovered = recovery.cycle_card(card, self._runner)
                        log.warning("Card recovery for %s: %s (attempt %d/%d)",
                                    name, recovered, self.capture.spent(name),
                                    self.capture.max_attempts)
        sink_mutes = {item["name"]: item["mute"]
                      for item in recovery._listing("sinks", self._runner)
                      if isinstance(item.get("name"), str) and isinstance(item.get("mute"), bool)} if sinks else {}
        for name, sink in sinks.items():
            if self._stop.is_set():
                return
            ptr, state = read_playback_status(sink["card"], sink["device"], sink["subdevice"])
            observed_at = time.monotonic() if now is None else now
            running = sink["running"] and sink_mutes.get(name) is False
            self.stall.observe(name, running, ptr, state, observed_at)
            if self.stall.just_stalled(name):
                log.warning("%s playback hardware is stalled; auto-recover %s",
                            name, self.auto_recover)
            if self.auto_recover and self.stall.should_recover(name, observed_at):
                if self._stop.is_set():
                    return
                self.stall.record_attempt(name, time.monotonic() if now is None else now)
                recovered = recycle_sink(name, self._runner)
                log.warning("Sink recovery for %s: %s (attempt %d/%d)",
                            name, recovered, self.stall.spent(name), self.stall.max_attempts)

    def _run(self):
        while not self._stop.is_set():
            try:
                self.check_once()
            except Exception:
                log.exception("Health monitor collection failed")
            self._stop.wait(CHECK_INTERVAL)

File: crates/openwave-core/tests/health.rs
use openwave_core::{
    health::*,
    model::{NodeIdentity, Observation, OperationError},
};
use serde_json::{Value, json};
use std::time::Duration;

fn t(seconds: u64) -> Duration {
    Duration::from_secs(seconds)
}
fn id(serial: &str) -> NodeIdentity {
    NodeIdentity {
        server_cookie: 42,
        object_serial: serial.into(),
    }
}
fn unknown<T>() -> Observation<T> {
    Observation::Unknown(OperationError::unavailable("collector unavailable"))
}
fn capture(serial: &str, count: u64, age: u64) -> Observation<CaptureSample> {
    Observation::Known(CaptureSample {
        identity: id(serial),
        running: true,
        muted: Observation::Known(false),
        xruns: Observation::Known(count),
        byte_age: Observation::Known(t(age)),
    })
}
fn sink(serial: &str, pointer: u64, state: PlaybackState) -> Observation<SinkSample> {
    Observation::Known(SinkSample {
        identity: id(serial),
        running: true,
        muted: Observation::Known(false),
        playback: Observation::Known(PlaybackStatus {
            hw_ptr: pointer,
            state,
        }),
    })
}

#[test]
fn sustained_xruns_require_two_windows_and_one_announcement() {
    let mut watch = GlitchWatch::default();
    assert!(!watch.observe("mic", &id("one"), 61994));
    assert_eq!(watch.last_delta("mic"), None);
    assert!(watch.observe("mic", &id("one"), 62044));
    assert!(!watch.glitching("mic"));
    assert!(watch.observe("mic", &id("one"), 62094));
    assert!(watch.glitching("mic") && watch.just_confirmed("mic"));
    watch.observe("mic", &id("one"), 62144);
    assert!(watch.glitching("mic") && !watch.just_confirmed("mic"));
    watch.observe("mic", &id("one"), 62167);
    assert!(!watch.glitching("mic"));
    watch.observe("mic", &id("one"), 62217);
    watch.observe("mic", &id("one"), 62218);
    assert!(!watch.glitching("mic"));
    watch.observe("mic", &id("two"), 90000);
    assert_eq!(watch.last_delta("mic"), None);
    watch.observe("mic", &id("two"), 0);
    assert_eq!(watch.last_delta("mic"), None);
    watch.pause("mic");
    watch.observe("mic", &id("two"), 200);
    assert!(!watch.glitching("mic"));
}

#[test]
fn no_data_and_xruns_spend_one_capped_cooled_budget() {
    let mut watch = CaptureWatch::default();
    assert!(
        watch
            .observe("mic", &capture("one", 0, 8), t(0))
            .recovery_due
    );
    watch.record_attempt("mic", t(0));
    assert!(!watch.observe("mic", &capture("one", 0, 0), t(10)).healthy);
    watch.observe("mic", &capture("one", 230, 0), t(20));
    let fault = watch.observe("mic", &capture("one", 460, 0), t(30));
    assert!(fault.glitching && !fault.recovery_due);
    assert!(
        watch
            .observe("mic", &capture("one", 690, 0), t(60))
            .recovery_due
    );
    watch.record_attempt("mic", t(60));
    assert!(
        !watch
            .observe("mic", &capture("one", 690, 100), t(300))
            .recovery_due
    );
    assert!(
        watch
            .observe("other", &capture("other", 0, 100), t(300))
            .recovery_due
    );
}

#[test]
fn unknown_muted_absent_and_recreated_capture_never_refill() {
    let mut watch = CaptureWatch::default();
    watch.record_attempt("mic", t(0));
    watch.record_attempt("mic", t(60));
    watch.observe("mic", &capture("one", 0, 0), t(100));
    watch.observe("mic", &capture("one", 0, 0), t(110));
    watch.observe("mic", &unknown(), t(400));
    watch.observe("mic", &capture("one", 0, 0), t(500));
    watch.observe("mic", &capture("one", 0, 0), t(510));
    let mut muted = capture("one", 0, 0);
    if let Observation::Known(sample) = &mut muted {
        sample.muted = Observation::Known(true);
    }
    watch.observe("mic", &muted, t(800));
    watch.observe("mic", &capture("one", 0, 0), t(900));
    watch.observe("mic", &capture("one", 0, 0), t(910));
    watch.observe("mic", &capture("two", 0, 0), t(1210));
    assert_eq!(watch.budget.spent("mic"), 2);
    assert!(
        !watch
            .observe("mic", &capture("two", 0, 0), t(1500))
            .recovery_due
    );
    assert!(
        !watch
            .observe("mic", &capture("two", 0, 9), t(1799))
            .recovery_due
    );
    watch.observe("mic", &capture("two", 0, 0), t(1800));
    watch.observe("mic", &capture("two", 0, 0), t(2100));
    assert!(
        watch
            .observe("mic", &capture("two", 0, 9), t(2101))
            .recovery_due
    );
}

#[test]
fn capture_counter_reset_and_missing_bytes_break_healthy_interval() {
    let mut watch = CaptureWatch::default();
    watch.record_attempt("mic", t(0));
    watch.record_attempt("mic", t(60));
    watch.observe("mic", &capture("one", 100, 0), t(100));
    watch.observe("mic", &capture("one", 100, 0), t(110));
    watch.observe("mic", &capture("one", 0, 0), t(410));
    assert_eq!(watch.budget.spent("mic"), 2);
    watch.observe("mic", &capture("one", 1, 0), t(500));
    let mut missing_bytes = capture("one", 1, 0);
    if let Observation::Known(sample) = &mut missing_bytes {
        sample.byte_age = unknown();
    }
    watch.observe("mic", &missing_bytes, t(800));
    watch.observe("mic", &capture("one", 1, 0), t(900));
    assert_eq!(watch.budget.spent("mic"), 2);
    watch.observe("mic", &capture("one", 1, 0), t(1200));
    assert_eq!(watch.budget.spent("mic"), 0);
}

#[test]
fn sink_stall_uses_running_pointer_and_xrun_not_audio_amplitude() {
    let mut watch = SinkStallWatch::default();
    assert!(!watch.observe("sink", &sink("one", 1000, PlaybackState::Running)));
    assert!(watch.observe("sink", &sink("one", 1000, PlaybackState::Running)));
    assert!(watch.just_stalled("sink") && watch.should_recover("sink", t(10)));
    watch.observe("sink", &sink("one", 1000, PlaybackState::Running));
    assert!(!watch.just_stalled("sink"));
    watch.record_attempt("sink", t(10));
    assert!(!watch.observe("sink", &sink("one", 0, PlaybackState::Running)));
    watch.observe("sink", &sink("one", 0, PlaybackState::Running));
    assert!(!watch.should_recover("sink", t(69)));
    assert!(watch.should_recover("sink", t(70)));
    watch.record_attempt("sink", t(70));
    assert!(watch.observe("sink", &sink("one", 0, PlaybackState::Xrun)));
    assert!(!watch.should_recover("sink", t(300)));
    assert!(watch.observe("other", &sink("other", 0, PlaybackState::Xrun)));
    assert!(watch.should_recover("other", t(300)));
    assert!(!watch.observe("sink", &unknown()));
    assert!(!watch.observe("sink", &sink("one", 1000, PlaybackState::Other)));
    let mut muted = sink("one", 1000, PlaybackState::Xrun);
    if let Observation::Known(sample) = &mut muted {
        sample.muted = Observation::Known(true);
    }
    assert!(!watch.observe("sink", &muted));
}

#[test]
fn only_six_advances_refill_output_not_reset_recreation_or_unknown() {
    let mut watch = SinkStallWatch::default();
    watch.record_attempt("sink", t(0));
    watch.record_attempt("sink", t(60));
    for pointer in 0..=5 {
        watch.observe("sink", &sink("one", pointer, PlaybackState::Running));
    }
    assert_eq!(watch.spent("sink"), 2);
    watch.observe("sink", &sink("one", 0, PlaybackState::Running));
    assert_eq!(watch.spent("sink"), 2);
    for pointer in 1..=5 {
        watch.observe("sink", &sink("one", pointer, PlaybackState::Running));
    }
    watch.observe("sink", &sink("two", 100, PlaybackState::Running));
    assert_eq!(watch.spent("sink"), 2);
    for pointer in 101..=105 {
        watch.observe("sink", &sink("two", pointer, PlaybackState::Running));
    }
    watch.observe("sink", &unknown());
    for pointer in 106..=111 {
        watch.observe("sink", &sink("two", pointer, PlaybackState::Running));
    }
    assert_eq!(watch.spent("sink"), 2);
    watch.observe("sink", &sink("two", 112, PlaybackState::Running));
    assert_eq!(watch.spent("sink"), 0);
    watch.observe("sink", &sink("two", 112, PlaybackState::Running));
    assert!(watch.should_recover("sink", t(1000)));
}

#[test]
fn failed_collection_and_valid_omission_pause_both_budgets() {
    let mut watch = HealthWatch::default();
    watch.capture.record_attempt("mic", t(0));
    watch.capture.record_attempt("mic", t(60));
    watch.sink.record_attempt("sink", t(0));
    watch.sink.record_attempt("sink", t(60));
    let frame = |pointer| {
        Observation::Known(HealthSamples {
            captures: [("mic".into(), capture("one", 0, 0))].into_iter().collect(),
            sinks: [("sink".into(), sink("one", pointer, PlaybackState::Running))]
                .into_iter()
                .collect(),
        })
    };
    watch.observe(&frame(0), t(100));
    watch.observe(&frame(1), t(110));
    assert!(watch.observe(&unknown(), t(500)).error.is_some());
    watch.observe(&frame(2), t(600));
    watch.observe(&frame(3), t(610));
    watch.observe(&Observation::Known(HealthSamples::default()), t(1000));
    watch.observe(&frame(4), t(1100));
    assert_eq!(watch.capture.budget.spent("mic"), 2);
    assert_eq!(watch.sink.spent("sink"), 2);
}

#[test]
fn capture_stall_boundary_absence_and_cooldown_are_conservative() {
    let mut watch = StallWatch::default();
    assert!(!watch.should_recover("mic", false, Some(t(999)), t(100)));
    assert!(!watch.should_recover("mic", true, None, t(100)));
    assert!(!watch.should_recover("mic", true, Some(t(7)), t(100)));
    assert!(watch.should_recover("mic", true, Some(t(8)), t(100)));
    watch.record_attempt("mic", t(100));
    assert!(!watch.can_recover("mic", t(99)));
    assert!(!watch.can_recover("mic", t(159)));
    assert!(watch.can_recover("mic", t(160)));
}

#[test]
fn collector_parsers_keep_last_iteration_and_reject_unknown_values() {
    let text = "S ID QUANT RATE WAIT BUSY W/Q B/Q ERR FORMAT NAME\nC 73 0 0 --- --- --- --- 0 dock\nR 73 0 0 0us 0us ??? ??? 30367 S24LE 1 48000 + dock\nR 199 0 0 0us 0us 0 0 5 F32P 1 0 + fx\nR 200 0 0 0us 0us 0 0 9 fx\n";
    let counts = parse_pw_top(text);
    assert_eq!(counts[&73].count, 30367);
    assert_eq!(counts[&199].count, 5);
    assert_eq!(counts[&200].count, 9);
    assert_eq!(counts.len(), 3);
    assert_eq!(
        parse_playback_status("state: RUNNING\nhw_ptr : 49000\n").unwrap(),
        PlaybackStatus {
            state: PlaybackState::Running,
            hw_ptr: 49000
        }
    );
    assert!(parse_playback_status("state: RUNNING\nhw_ptr: unknown").is_err());
    assert!(parse_mutes(&json!([{"name":"mic", "mute":null}])).is_err());
    assert!(
        parse_mutes(&json!([{"name":"mic", "mute":false}, {"name":"mic", "mute":true}])).is_err()
    );
    assert_eq!(
        parse_mutes(&json!([{"name":"mic", "mute":false}])).unwrap()["mic"],
        false
    );
}

fn graph_node(id: u32, name: &str, class: &str) -> Value {
    json!({"id":id,"type":"PipeWire:Interface:Node","info":{"state":"running", "props":{"node.name":name,"media.class":class,"object.serial":id+100,"api.alsa.pcm.card":0,"api.alsa.pcm.device":1}}})
}
fn graph_link(source: u32, target: u32) -> Value {
    json!({"type":"PipeWire:Interface:Link","info":{"output-node-id":source,"input-node-id":target}})
}
#[test]
fn output_watch_follows_actual_links_and_rejects_malformed_graph() {
    let mut route = graph_node(5, "openwave_loop_output_personal", "Stream/Output/Audio");
    route["info"]["props"]["target.object"] = json!("alsa_output.unrelated");
    let mut graph = json!([
        {"id":0,"type":"PipeWire:Interface:Core","info":{"cookie":42}},
        graph_node(1,"alsa_input.usb-Elgato_Systems_Elgato_XLR_Dock_SERIAL-00.mono-fallback","Audio/Source"),
        graph_node(2,"alsa_output.linked","Audio/Sink"),
        graph_node(3,"alsa_output.unrelated","Audio/Sink"),
        graph_node(4,"openwave_loop_output_personal_cap","Stream/Input/Audio"),
        route, graph_node(6,"unrelated_player","Stream/Output/Audio"),
        graph_link(5,2), graph_link(4,3), graph_link(6,3)
    ]);
    let parsed = parse_health_graph(&graph).unwrap();
    assert_eq!(
        parsed.sinks.keys().collect::<Vec<_>>(),
        [&"alsa_output.linked".to_string()]
    );
    assert_eq!(
        parsed.sinks["alsa_output.linked"].playback,
        AlsaPlayback {
            card: 0,
            device: 1,
            subdevice: 0
        }
    );
    assert_eq!(parsed.captures.values().next().unwrap().identity, id("101"));
    graph[2]["info"]["props"] = json!([]);
    assert!(parse_health_graph(&graph).is_err());
    graph[2] = graph_node(2, "alsa_output.linked", "Audio/Sink");
    graph[0]["info"]["cookie"] = Value::Null;
    assert!(parse_health_graph(&graph).is_err());
}
